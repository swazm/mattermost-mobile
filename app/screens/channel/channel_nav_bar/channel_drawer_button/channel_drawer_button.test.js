// Copyright (c) 2015-present SWAZM, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';
import {shallow} from 'enzyme';
import NotificationsIOS from 'react-native-notifications';

import Preferences from 'mattermost-redux/constants/preferences';

import Badge from 'app/components/badge';
import PushNotification from 'app/push_notifications/push_notifications.ios';

import ChannelDrawerButton from './channel_drawer_button';

jest.mock('react-native-notifications', () => {
    let badgesCount = 0;
    let deliveredNotifications = {};

    return {
        getBadgesCount: jest.fn((callback) => callback(badgesCount)),
        setBadgesCount: jest.fn((count) => {
            badgesCount = count;
        }),
        addEventListener: jest.fn(),
        setDeliveredNotifications: jest.fn((notifications) => {
            deliveredNotifications = notifications;
        }),
        getDeliveredNotifications: jest.fn(async (callback) => {
            await callback(deliveredNotifications);
        }),
        removeDeliveredNotifications: jest.fn((ids) => {
            deliveredNotifications = deliveredNotifications.filter((n) => !ids.includes(n.identifier));
        }),
        cancelAllLocalNotifications: jest.fn(),
        NotificationAction: jest.fn(),
        NotificationCategory: jest.fn(),
    };
});

describe('ChannelDrawerButton', () => {
    const baseProps = {
        openDrawer: jest.fn(),
        badgeCount: 0,
        theme: Preferences.THEMES.default,
        visible: false,
    };

    afterEach(() => NotificationsIOS.setBadgesCount(0));

    test('should match, full snapshot', () => {
        const wrapper = shallow(
            <ChannelDrawerButton {...baseProps}/>
        );

        // no badge to show
        expect(wrapper.getElement()).toMatchSnapshot();
        expect(wrapper.find(Badge).length).toEqual(0);

        // badge should render
        wrapper.setProps({badgeCount: 1, visible: true});
        expect(wrapper.getElement()).toMatchSnapshot();
        expect(wrapper.find(Badge).length).toEqual(1);
    });

    test('should not set app icon badge on mount', () => {
        const setApplicationIconBadgeNumber = jest.spyOn(PushNotification, 'setApplicationIconBadgeNumber');
        const props = {
            ...baseProps,
            badgeCount: 0,
        };

        shallow(
            <ChannelDrawerButton {...props}/>
        );
        expect(setApplicationIconBadgeNumber).not.toBeCalled();
        NotificationsIOS.getBadgesCount((count) => expect(count).toBe(0));
    });

    test('should set app icon badge on mount', () => {
        const setApplicationIconBadgeNumber = jest.spyOn(PushNotification, 'setApplicationIconBadgeNumber');
        const props = {
            ...baseProps,
            badgeCount: 1,
        };

        shallow(
            <ChannelDrawerButton {...props}/>
        );
        expect(setApplicationIconBadgeNumber).toHaveBeenCalledTimes(1);
        NotificationsIOS.getBadgesCount((count) => expect(count).toBe(1));
    });

    test('should set app icon badge update', () => {
        const setApplicationIconBadgeNumber = jest.spyOn(PushNotification, 'setApplicationIconBadgeNumber');
        const props = {
            ...baseProps,
            badgeCount: 0,
        };

        const wrapper = shallow(
            <ChannelDrawerButton {...props}/>
        );
        NotificationsIOS.getBadgesCount((count) => expect(count).toBe(0));

        wrapper.setProps({badgeCount: 2});
        expect(setApplicationIconBadgeNumber).toHaveBeenCalledTimes(1);
        expect(setApplicationIconBadgeNumber).toHaveBeenCalledWith(2);
        NotificationsIOS.getBadgesCount((count) => expect(count).toBe(2));
    });

    test('should set remove icon badge on update', () => {
        const setApplicationIconBadgeNumber = jest.spyOn(PushNotification, 'setApplicationIconBadgeNumber');
        const props = {
            ...baseProps,
            badgeCount: 0,
        };

        const wrapper = shallow(
            <ChannelDrawerButton {...props}/>
        );
        wrapper.setProps({badgeCount: 2});
        expect(setApplicationIconBadgeNumber).toHaveBeenCalledWith(2);
        NotificationsIOS.getBadgesCount((count) => expect(count).toBe(2));

        wrapper.setProps({badgeCount: -1});
        expect(setApplicationIconBadgeNumber).toHaveBeenCalledWith(-1);
        NotificationsIOS.getBadgesCount((count) => expect(count).toBe(0));
    });
});
