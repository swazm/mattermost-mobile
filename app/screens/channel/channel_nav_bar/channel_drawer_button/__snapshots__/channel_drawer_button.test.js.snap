// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChannelDrawerButton should match, full snapshot 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  onPress={[Function]}
  style={
    Object {
      "width": 10,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flex": 1,
          "flexDirection": "column",
          "justifyContent": "center",
          "paddingHorizontal": 10,
        },
      ]
    }
  >
    <View>
      <View
        style={
          Object {
            "height": 25,
          }
        }
      />
    </View>
  </View>
</TouchableOpacity>
`;

exports[`ChannelDrawerButton should match, full snapshot 2`] = `
<TouchableOpacity
  activeOpacity={0.2}
  onPress={[Function]}
  style={
    Object {
      "width": 55,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flex": 1,
          "flexDirection": "column",
          "justifyContent": "center",
          "paddingHorizontal": 10,
        },
      ]
    }
  >
    <View>
      <Icon
        allowFontScaling={false}
        color="#ffffff"
        name="md-menu"
        size={25}
      />
      <Badge
        containerStyle={
          Object {
            "borderColor": "#1153ab",
            "borderRadius": 14,
            "borderWidth": 2,
            "position": "absolute",
            "right": -14,
            "top": -7,
          }
        }
        count={1}
        countStyle={
          Object {
            "color": "#145dbf",
            "fontSize": 10,
          }
        }
        extraPaddingHorizontal={10}
        minHeight={20}
        minWidth={19}
        onPress={[Function]}
        style={
          Object {
            "backgroundColor": undefined,
            "height": 19,
            "padding": 3,
          }
        }
      />
    </View>
  </View>
</TouchableOpacity>
`;
